src/Akka.Streams.Kafka/Messages/CommittableOffsetBatch.cs

```diff
@@ -0,0 +1,152 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading.Tasks;
+using Akka.Streams.Kafka.Extensions;
+using Akka.Streams.Kafka.Stages.Consumers;
+
+namespace Akka.Streams.Kafka.Messages
+{
+    /// <summary>
+    /// Stores committable offsets batch and allows to commit them with <see cref="Commit"/> method
+    /// </summary>
+    internal sealed class CommittableOffsetBatch : ICommittableOffsetBatch
+    {
+        /// <summary>
+        /// CommittableOffsetBatch
+        /// </summary>
+        public CommittableOffsetBatch(IImmutableDictionary<GroupTopicPartition, OffsetAndMetadata> offsetsAndMetadata, 
+                                      IImmutableDictionary<string, IInternalCommitter> committers, 
+                                      long batchSize)
+        {
+            OffsetsAndMetadata = offsetsAndMetadata;
+            Committers = committers;
+            BatchSize = batchSize;
+        }
+
+        /// <inheritdoc />
+        public long BatchSize { get; }
+        
+        /// <inheritdoc />
+        public IImmutableSet<GroupTopicPartitionOffset> Offsets => OffsetsAndMetadata.Select(o => new GroupTopicPartitionOffset(o.Key, o.Value.Offset)).ToImmutableHashSet();
+        
+        /// <summary>
+        /// Committers
+        /// </summary>
+        public IImmutableDictionary<string, IInternalCommitter> Committers { get; }
+        
+        /// <summary>
+        /// Offsets and metadata
+        /// </summary>
+        public IImmutableDictionary<GroupTopicPartition, OffsetAndMetadata> OffsetsAndMetadata { get; }
+        
+        /// <summary>
+        /// Create empty offset batch
+        /// </summary>
+        public static ICommittableOffsetBatch Empty => new CommittableOffsetBatch(ImmutableDictionary<GroupTopicPartition, OffsetAndMetadata>.Empty, 
+                                                                                  ImmutableDictionary<string, IInternalCommitter>.Empty, 
+                                                                                  0);
+        
+        /// <summary>
+        /// Create an offset batch out of a first offsets.
+        /// </summary>
+        public static ICommittableOffsetBatch Create(ICommittableOffset offset) => Empty.Updated(offset);
+        /// <summary>
+        /// Create an offset batch out of a list of offsets.
+        /// </summary>
+        public static ICommittableOffsetBatch Create(IEnumerable<ICommittable> offsets)
+        {
+            return offsets.Aggregate(Empty, (batch, offset) => batch.Updated(offset));
+        }
+
+        /// <inheritdoc />
+        public async Task Commit()
+        {
+            if (Offsets.IsEmpty() || Committers.IsEmpty())
+                return;
+
+            await Committers.First().Value.Commit(this);
```

Comment:
Do all committers need to commit here? Why not?

Reply:
First of all, here is their code in alpakka:
```scala
if (offsets.isEmpty)
  Future.successful(Done)
else {
  committers.head._2.commit(this)
}
```

And why only one committer makes commit: that is because in committer's implementation it takes all `Committers` of this batch object, finds appropriate committer for each offsets group (you can see that `Committers `are grouped my groupId) and then commits them all. So internally multiple committers may be used. Here is a `Commit `implementation:

```csharp
public async Task Commit(ICommittableOffsetBatch batch)
{
	if (!(batch is CommittableOffsetBatch batchImpl))
		throw new ArgumentException($"Unknown CommittableOffsetBatch, got {batch.GetType().FullName}, but expected {nameof(CommittableOffsetBatch)}");
	
	await Task.WhenAll(batchImpl.OffsetsAndMetadata.GroupBy(o => o.Key.GroupId).Select(group =>
	{
		if (!batchImpl.Committers.TryGetValue(group.Key, out var committer))
			throw new IllegalStateException($"Unknown committer, got groupId = {group.Key}");

		var offsets = group.Select(offset => new GroupTopicPartitionOffset(offset.Key, offset.Value.Offset)).ToImmutableList();
		return committer.Commit(offsets);
	}));
}
```

Little bit messy, and the only place where this  `Commit(ICommittableOffsetBatch batch)` is used is that line you have a question about. So I could remove this method from committer and make committers iterating more explicit, right there. This will also allow make `Committers` private field, which seems convenient. What do you think?

Reply:
This is a great response - no worries. This implementation is fine and answers my question.